plugin/evm/vm_test.go

```diff
@@ -810,6 +817,192 @@ func TestIssueAtomicTxs(t *testing.T) {
+		t.Fatalf("Expected status of built block to be %s, but found %s", choices.Processing, status)
+	}
+
+	if err := vm.SetPreference(context.Background(), blk.ID()); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := blk.Accept(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	if status := blk.Status(); status != choices.Accepted {
+		t.Fatalf("Expected status of accepted block to be %s, but found %s", choices.Accepted, status)
+	}
+
+	if lastAcceptedID, err := vm.LastAccepted(context.Background()); err != nil {
+		t.Fatal(err)
+	} else if lastAcceptedID != blk.ID() {
+		t.Fatalf("Expected last accepted blockID to be the accepted block: %s, but found %s", blk.ID(), lastAcceptedID)
+	}
+
+	// The same atomic transaction is included in multiple processing blocks.
+	// It gets dropped in one block and is still processing in the other.
+	// getAtomicTx should return status Processing
+	exportTx, err := vm.newExportTx(vm.ctx.AVAXAssetID, 1, vm.ctx.XChainID, testShortIDAddrs[0], initialBaseFee, []*secp256k1.PrivateKey{testKeys[0]})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = vm.mempool.AddLocalTx(exportTx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	<-issuer
+
+	blk2, err := vm.BuildBlock(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = blk2.Verify(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := vm.SetPreference(context.Background(), blk2.ID()); err != nil {
+		t.Fatal(err)
+	}
+
+	err = vm.mempool.ForceAddTx(exportTx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	<-issuer
+
+	// new blk also picks up the same atomic tx, fails block building, and should discard the atomic tx
+	_, err = vm.BuildBlock(context.Background())
+	if err == nil {
+		t.Fatalf("this block should have failed verification since it holds the same tx as the preferred block")
+	}
+
+	if status := blk2.Status(); status != choices.Processing {
+		t.Fatalf("Expected status of accepted block to be %s, but found %s", choices.Processing, status)
+	}
+
+	// the atomic tx should be safe in blk2 for now
+	// and getAtomicTx should report as processing even though it got discarded
+	exportTxFetched, status, height, err := vm.getAtomicTx(exportTx.ID())
+	assert.NoError(t, err)
+	assert.Equal(t, Processing, status)
+	assert.Equal(t, uint64(2), height, "expected height of export tx to be 2")
+	assert.Equal(t, exportTxFetched.ID(), exportTx.ID(), "expected ID of fetched export tx to match original txID")
+
+	if err := blk2.Accept(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	if status := blk2.Status(); status != choices.Accepted {
+		t.Fatalf("Expected status of accepted block to be %s, but found %s", choices.Accepted, status)
+	}
+
+	if lastAcceptedID, err := vm.LastAccepted(context.Background()); err != nil {
+		t.Fatal(err)
+	} else if lastAcceptedID != blk2.ID() {
+		t.Fatalf("Expected last accepted blockID to be the accepted block: %s, but found %s", blk2.ID(), lastAcceptedID)
```

Comment:
can we add a check that `exportTx` is marked as accepted with a correctly indexed height here?

Reply:
yup